### PR TITLE
fix(upgrade): use correct attribute name for upgraded component's binings

### LINKED ISSUE #8856: UpgradeAdapter sets incorrect keys in ng1 directive controller
- **I'm submitting a bug report**

**Current behavior**
If ng1 directive controller has properties bound using bindToController and if the instance variable name is different from the attribute being read, then while using it through UpgradeAdapter, the controller gets all those properties with the attribute names and not the instance variable names.
for e.g.

```js
controller: function () {
  this.first;
  this.second;
  this.third;
},
bindToController: {
  first: '<firstValue',
  second: '=?secondValue',
  third: '=?'
},
controllerAs: 'vm'
```

The when you use the directive upgraded from UpgradeAdapter, then upon inspecting the controller these show up as:

```js
vm.firstValue
vm.?secondValue
vm.?
```

**Expected/desired behavior**
The values in controller should be populated in the variables first, second and third.
- **If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem** via
  Here's a plunkr to reproduce/see this issue.
  https://plnkr.co/edit/0FCNd7bQ9zM3jKTViD2t?p=preview
  If you notice i use the upgraded directive as:
  <ng1-directive [first]="'one'" [second]="'two'">
  And i have printed all the controller variables in the output and they end up as:
  First: 
  Second: 
  FirstValue: one
  Entire Controller: {"?secondValue":"two","firstValue":"one"}
- **What is the motivation / use case for changing the behavior?**
  Because of this bug, directives which have such bindings (which are genuinely required in some cases) do not work as expected. And frankly this is one of the numerous bugs we are coming across while using the UpgradeAdapter, using which is a pretty frustrating experience. It definitely does not work as it's expected to as per the guide.
- **Please tell us about your environment:**
- **Angular version:** 2.0.0-beta.17 
- **Browser:** [all | Chrome XX | Firefox XX | IE XX | Safari XX | Mobile Chrome XX | Android X.X Web Browser | iOS XX Safari | iOS XX UIWebView | iOS XX WKWebView ]

### PR DESCRIPTION
**Pull request by gkalpak**
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] ~~Docs have been added / updated (for bug fixes / features)~~

**What kind of change does this PR introduce?** (check one with "x")
```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

**What is the current behavior?** (You can also link to an open issue here)
When using a different property/attribute name for an upgraded component's binding (e.g. `bindings: {propName: '<attrName'}`), the property and attribute names are swapped (e.g. using `attrName` as the property name and `propName` as the attribute name). This results in unexpected behavior.
This only affects `upgrade/dynamic`. `upgrade/static` works correctly already.

**What is the new behavior?**
The correct names for properties and attributes are used.

**Does this PR introduce a breaking change?** (check one with "x")
```
[ ] Yes
[x] No **(but any apps relying on the previous, broken behavior will be broken)**
```

**Other information**:
Fixes #8856.

### CODE REVIEW

File: packages/upgrade/src/dynamic/upgrade_ng1_adapter.ts [hunk 1]
```diff
@@ -99,49 +99,50 @@ export class UpgradeNg1ComponentAdapterBuilder {
...
-              throw new Error(
-                  `Unexpected mapping '${type}' in '${json}' in '${this.name}' directive.`);
-          }
+      Object.keys(context).forEach(propName => {
+        const definition = context[propName];
+        const bindingType = definition.charAt(0);
+        const bindingOptions = definition.charAt(1);
+        const attrName = definition.substring(bindingOptions === '?' ? 2 : 1) || propName;
+
+        // QUESTION: What about `=*`? Ignore? Throw? Support?
```
Comment (petebacondarwin) [inline packages/upgrade/src/dynamic/upgrade_ng1_adapter.ts]: Ignore, given that this is upgrade "dynamic" which is not our focus of attention going forward.

File: packages/upgrade/src/dynamic/upgrade_ng1_adapter.ts [hunk 2]
```diff
@@ -99,49 +99,50 @@ export class UpgradeNg1ComponentAdapterBuilder {
...
-            default:
-              let json = JSON.stringify(context);
-              throw new Error(
-                  `Unexpected mapping '${type}' in '${json}' in '${this.name}' directive.`);
-          }
+      Object.keys(context).forEach(propName => {
+        const definition = context[propName];
+        const bindingType = definition.charAt(0);
+        const bindingOptions = definition.charAt(1);
+        const attrName = definition.substring(bindingOptions === '?' ? 2 : 1) || propName;
```
Comment (petebacondarwin) [inline packages/upgrade/src/dynamic/upgrade_ng1_adapter.ts]: I have a feeling that your supreme regex skills could make this code a lot shorter.

**Review by petebacondarwin — APPROVED** [review]

**Comment by petebacondarwin** [comment]
LGTM - although there is a conflict, which made me realise... have you checked that your new code supports "nullability"? See https://github.com/angular/angular/commit/01d93f3af8c427dd3624bc50cc1a51d6f5bec906#diff-a7126213c5a3fc449c88971bb55add43

File: packages/upgrade/src/dynamic/upgrade_ng1_adapter.ts (hunk 1 above)
Comment (gkalpak) [inline packages/upgrade/src/dynamic/upgrade_ng1_adapter.ts]: Actually, we have the same issue in `upgrade/static`. I have copied part of the bindngs implementation (including this comment) from the static version :grin:

File: packages/upgrade/src/dynamic/upgrade_ng1_adapter.ts (hunk 2 above)
Comment (gkalpak) [inline packages/upgrade/src/dynamic/upgrade_ng1_adapter.ts]: True story 😛 These 4 lines could probably be:

```ts
[, bindingType, attrName = propName] = /^(.)\??(.+)?$/.exec(context[propName]);
```

But I copied this from `static` and it is probably more readable for people with less supreme regex skills 😛

**Comment by gkalpak** [comment]
I didn't think about `strictNullChecks`. Good catch.
Rebased on master, fixed `strictNullChecks` with ab9c29136, then squashed.

**Review by petebacondarwin — APPROVED** [review]
still LGTM

**Comment by gkalpak** [comment]
I fixed the broken build. Apparently, there was another `strictNullChecks` violation that `gulp lint` didn't catch:

```diff
-const ng1Controller = angular.element(ng1).controller('ng1');
+const ng1Controller = angular.element(ng1).controller !('ng1');
```

**Pull request closed without merging**